storage-proofs/src/circuit/rational_post.rs

```diff
@@ -0,0 +1,377 @@
+            challenges_count,
+        };
+
+        let data1: Vec<u8> = (0..32)
+            .flat_map(|_| fr_into_bytes::<Bls12>(&rng.gen()))
+            .collect();
+        let data2: Vec<u8> = (0..32)
+            .flat_map(|_| fr_into_bytes::<Bls12>(&rng.gen()))
+            .collect();
+
+        let graph1 = BucketGraph::<PedersenHasher>::new(32, 5, 0, new_seed());
+        let tree1 = graph1.merkle_tree(data1.as_slice()).unwrap();
+
+        let graph2 = BucketGraph::<PedersenHasher>::new(32, 5, 0, new_seed());
+        let tree2 = graph2.merkle_tree(data2.as_slice()).unwrap();
+
+        let faults = vec![];
+        let seed = (0..32).map(|_| rng.gen()).collect::<Vec<u8>>();
+        let challenges = derive_challenges(challenges_count, sector_size, 2, &seed, &faults);
+        let commitments_raw = vec![tree1.root(), tree2.root()];
+        let commitments: Vec<_> = challenges
+            .iter()
+            .map(|c| commitments_raw[c.sector as usize])
+            .collect();
+
+        let pub_inputs = rational_post::PublicInputs {
+            challenges: &challenges,
+            faults: &faults,
+            commitments: &commitments,
+        };
+
+        let priv_inputs = rational_post::PrivateInputs::<PedersenHasher> {
+            trees: &[&tree1, &tree2],
+        };
+
+        let proof = RationalPoSt::<PedersenHasher>::prove(&pub_params, &pub_inputs, &priv_inputs)
+            .expect("proving failed");
+
+        let is_valid = RationalPoSt::<PedersenHasher>::verify(&pub_params, &pub_inputs, &proof)
+            .expect("verification failed");
+        assert!(is_valid);
+
+        // actual circuit test
+
+        let paths: Vec<_> = proof
+            .paths()
+            .iter()
+            .map(|p| {
+                p.iter()
+                    .map(|v| Some((v.0.into(), v.1)))
+                    .collect::<Vec<_>>()
+            })
+            .collect();
+        let leafs: Vec<_> = proof.leafs().iter().map(|l| Some((**l).into())).collect();
+
+        let mut cs = TestConstraintSystem::<Bls12>::new();
+
+        let instance = RationalPoStCircuit::<_, PedersenHasher> {
+            params,
+            leafs,
+            paths,
+            commitments: commitments.into_iter().map(|c| Some(c.into())).collect(),
+            _h: PhantomData,
+        };
+
+        instance
+            .synthesize(&mut cs)
+            .expect("failed to synthesize circuit");
+
+        assert!(cs.is_satisfied(), "constraints not satisfied");
+
+        assert_eq!(cs.num_inputs(), 5, "wrong number of inputs");
+        assert_eq!(cs.num_constraints(), 13746, "wrong number of constraints");
+        assert_eq!(cs.get_input(0, "ONE"), Fr::one());
+    }
+
+    #[ignore] // Slow test – run only when compiled for release.
+    #[test]
+    fn rational_post_test_compound() {
+        let params = &JubjubBls12::new();
+        let rng = &mut XorShiftRng::from_seed([0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
```

Comment:
QUESTION: What are these magic values? Where do they come from?

Reply:
we copy them from one test to the other, it's just some seed so we have reproducibility between tests,

Reply:
> we copy them from one test to the other

Will you factor these magic values into a constant that can be shared? Looking through the codebase, I found 49 instances of `[0x3dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]`.

Reply:
not in this pr

Reply:
https://github.com/filecoin-project/rust-fil-proofs/issues/838